KAFKA-17928: Make remote log manager thread-pool configs dynamic

kamalcph opened the pull request:
Co-authored-by: Peter Lee <peterxcli@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

peterxcli committed:
KAFKA-17928: Make remote log manager thread-pool configs dynamic

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

kamalcph commented:
cc @peterxcli

peterxcli reviewed (approved):
LGTM, thanks

satishd commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -2151,31 +2167,30 @@ RLMTaskWithFuture followerTask(TopicIdPartition partition) {
-                }
-            });
-
+            threadPool.setThreadFactory(ThreadUtils.createThreadFactory(threadNamePattern, true));
```
This is a good cleanup.

satishd commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1167,6 +1167,22 @@ class DynamicRemoteLogConfig(server: KafkaBroker) extends BrokerReconfigurable w
+        val oldValue = server.config.getInt(k)
+        if (newValue != oldValue) {
+          val errorMsg = s"Dynamic thread count update validation failed for $k=$v"
+          if (newValue <= 0)
```
The configuration can be updated to its default value of -1 through a static configuration update.

Is this an intended change that does not allow updating it to the default value (-1) through dynamic configs?

peterxcli replied:
I guess the only scenario where the user might want to set it back is if they want all three(maybe two😆) thread pools to follow the changes to `REMOTE_LOG_MANAGER_THREAD_POOL_SIZE_PROP` simultaneously.

Therefore, if we allow the user to do this, I think we should also make `REMOTE_LOG_MANAGER_THREAD_POOL_SIZE_PROP` dynamically configurable.

What are your thoughts on this?

kamalcph replied:
> The configuration can be updated to its default value of -1 through a static configuration update.

This is allowed. Only the newValue which is being dynamically updated to -1 is not allowed. 

> Is this an intended change that does not allow updating it to the default value (-1) through dynamic configs?

Yes, we made those thread-pools default to -1 to derive the value from `REMOTE_LOG_MANAGER_THREAD_POOL_SIZE_PROP`. Once those properties are configured to be a non-negative number, then we expect the user to provide subsequent valid value when updating it dynamically.

satishd replied:
I understand the default behavior, I suggested keeping -1 for falling back to the default value in the original [PR](https://github.com/apache/kafka/pull/17499/#discussion_r1804119567). 

I am fine with the approach of not allowing back to setting default as -1 for the dynamic configuration.  Please add a java doc comment and also add a clear message in the exception for user.

kamalcph replied:
Updated the javadoc. Keeping the thread pool default value as -1 is confusing. As part of [KIP-1030](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1030%3A+Change+constraints+and+default+values+for+various+configurations), can we change the default value to 10 and document it in the upgrade notes?

WDYT? cc @showuon

kamalcph replied:
We can take this discussion separately to unblock this PR.

kamalcph replied:
Initiated a discussion to change the default value to 10 in [KIP-1030](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1030%3A+Change+constraints+and+default+values+for+various+configurations) thread.

kamalcph replied:
Removed the validation for copier and expiration thread pool. It follows the default validation set in ConfigDef (Value can be -1 or greater than 0). With this, the static and dynamic validations are same.

satishd reviewed (commented):
Thanks @kamalcph @peterxcli for the PR, left a comment.

kamalcph commented:
@satishd 

Addressed your review comment. PTAL.

satishd commented:
@kamalcph As we discussed, let us revisit this PR once KIP-1030 is finalised.

kamalcph committed:
Disallow configuring -1 for copier and expiration thread pools

To provide clear configuration and reduce the need for manual tuning, the following changes are proposed:

- Change the default value of copier and expiration thread pool size from -1 to 10. This will also ensure smooth validation (step increments x/2 to x*2) of thread-count dynamic config change.
- Change the default value of remote.log.manager.thread.pool.size from 10 to 2, as the task is light-weight to fetch the highest-uploaded remote offset for the follower partitions.
- We can deprecate the logic that refers the value from another thread-pool (transitive config dependency). The thread-pool configured values become independent of each other.

kamalcph commented:
@satishd @divijvaidya @christo

The PR is ready for review. PTAL.

satishd commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -2152,31 +2168,30 @@ RLMTaskWithFuture followerTask(TopicIdPartition partition) {
-                }
-            });
-
+            threadPool.setThreadFactory(ThreadUtils.createThreadFactory(threadNamePattern, true));
```
One minor difference is that `KafkaThread` sets the `UncaughtExceptionHandler` like below:
```
setUncaughtExceptionHandler((t, e) -> log.error("Uncaught exception in thread '{}':", name, e));
```

kamalcph replied:
Good catch!

satishd reviewed (commented):
Thanks @kamalcph for addressing the review comments, overall LGTM.

kamalcph committed:
Addressed the review comments

chia7712 commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -202,6 +202,81 @@ class DynamicBrokerConfigTest {
+
+    val props3 = new Properties()
+    props3.put(RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP, "-1")
+    val err3 = assertThrows(classOf[ConfigException], () => config.dynamicConfig.validate(props, perBrokerConfig = true))
```
It should be `props3` rather than `props`. I noticed this when reviewing #22349 :)

we will file a small patch for it

FrankYang0529 replied:
https://github.com/apache/kafka/pull/22352
